src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs

```diff
@@ -1405,6 +1428,10 @@ internal void Stop()
                 copyStack = _stack.ToArray();
             }
 
+            // We don't want to stop when a pipeline is in a critical region.
+            // Typically it will be in the middle of command disposal.
+            AwaitPipelineCriticalSection();
```

Comment:
> Here is where I'd expect a lock to be taken, and later released further down in the call

It's only meant to stop the stopper from cancelling `cleanup` when it's being invoked during normal execution (e.g. just at the end of the pipeline's lifetime or an exception is thrown).  I don't think it's needed if the pipeline stopper is *initiating* the stop since the stopper is already busy and subsequent calls will be ignored.

Reply:
StopProcessing() currently doesn't prevent other StopProcessing() calls from occurring on a different thread. Do we want to change that?